src/Metadata/MetadataParser.php

```diff
@@ -0,0 +1,168 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\Image\Metadata;
+
+use Contao\Image\Exception\InvalidArgumentException;
+use Contao\Image\Exception\InvalidImageContainerException;
+use Contao\Image\Exception\InvalidImageMetadataException;
+
+/**
+ * @final
+ */
+class MetadataParser
+{
+    /**
+     * @var list<AbstractContainer>
+     */
+    private $containers;
+
+    /**
+     * @var array<string,AbstractFormat>
+     */
+    private $formats;
+
+    /**
+     * @param list<AbstractFormat>    $formats
+     * @param list<AbstractContainer> $containers
+     */
+    public function __construct(array $formats = [], array $containers = [])
```

Comment:
```suggestion
    public function __construct(iterable $formats = [], iterable $containers = [])
```

Reply:
The assignment `$this->containers = …` then needs to be adjusted as well to a `foreach` or the classic `is_array` check +`iterator_to_array`.